linera-base/src/abi.rs

```diff
@@ -70,3 +72,24 @@ where
     type Query = <<A as WithServiceAbi>::Abi as ServiceAbi>::Query;
     type QueryResponse = <<A as WithServiceAbi>::Abi as ServiceAbi>::QueryResponse;
 }
+
+/// The type of a query for an oracle.
+pub trait OracleQuery: Send + Sync + Debug + 'static {
+    /// The oracle's ID.
+    const ID: OracleId;
+
+    /// The response type of the oracle.
+    type Response: Send + Sync + Debug + 'static;
+
+    /// Error serializing the query.
+    type SerializationError: Debug + Display + 'static;
+
+    /// Error deserializing the response.
+    type DeserializationError: Debug + Display + 'static;
+
+    /// Serializes the query.
+    fn serialize(&self) -> Result<Vec<u8>, Self::SerializationError>;
+
+    /// Deserializes the response.
+    fn deserialize(response: Vec<u8>) -> Result<Self::Response, Self::DeserializationError>;
```

Comment:
Why not use a `response: &[u8]`?

Reply:
I think after it came through the Wasm boundary we always own the vector. (Correct me if I'm wrong, @jvff!)
So for oracles that _want_ to return just bytes anyway, and don't do any serialization, it's nice if they don't have to clone it.

Reply:
My view was different:
* When serializing, or deserializing you never need to own the input. So, for example, the `serialize(&self)` function that you wrote is taking a reference.
* Even if we own the vector in Wasm land, we may want to use it for something else.

Anyway, it is a minor point.

Reply:
OK, I made it a reference. We can always change it later, if needed.